README.md
# Generate vCard

[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![npm-version](https://img.shields.io/npm/v/generate-vcard?logo=npm)](https://www.npmjs.com/package/generate-vcard)
[![coverage](https://img.shields.io/codecov/c/github/1nVitr0/lib-generate-vcard?logo=codecov&token=D1VD9GHM8B)](https://codecov.io/gh/1nVitr0/lib-generate-vcard)

A simple library for vCard generation compatible with [RFC 6350](https://tools.ietf.org/html/rfc6350) and additions defined in [RFC 6474](https://tools.ietf.org/html/rfc6474), [RFC 6715](https://tools.ietf.org/html/rfc6715) and [RFC 8605](https://tools.ietf.org/html/rfc8605).

The library also supports some experimental drafts for additional properties and parameters.
As they are not part of the official IANA spec, they will be prefixed with `X-` in the generated vCard.
The supported additions are some social properties as defined in [draft-george-vcarddav-vcard-extension-02](https://tools.ietf.org/id/draft-george-vcarddav-vcard-extension-02.html),
the service type parameter for `IMPP` properties as defined in [draft-daboo-vcard-service-type](https://datatracker.ietf.org/doc/html/draft-daboo-vcard-service-type)
as well as the `X-ABLabel` property used by GMail and iOs for URL labelling, as described [here](https://alessandrorossini.org/the-sad-story-of-the-vcard-format-and-its-lack-of-interoperability/).

## Documentation

For more documentation see [https://1nvitr0.github.io/lib-generate-vcard/](https://1nvitr0.github.io/lib-generate-vcard/modules/generate_vcard.html).

## Installation

```bash
npm install --save generate-vcard
```

## Usage

```javascript
import generateVCard, { Kind } from "generate-vcard";

const vCard = generateVCard({
  kind: Kind.Individual,
  fullName: "Jane Doe",
  name: { honorificPrefix: "Dr.", familyName: "Doe", givenName: "Jane"  },
  socialProfile: [
    { value: "https://instagram.com/", parameters: { type: "instagram" } },
    { value: "https://twitter.com/", parameters: { type: "twitter" } },
  ]
});
```

> ```text
> BEGIN:VCARD
> VERSION:4.0
> KIND:individual
> FN:Jane Doe
> N:Doe;Jane;;Dr.;
> X-SOCIALPROFILE;TYPE=instagram:https://instagram.com/
> X-SOCIALPROFILE;TYPE=twitter:https://twitter.com/
> END:VCARD
> ```

Alternative usages:

```javascript
import { VCard, Kind } from "generate-vcard";

const vCard = new VCard(Kind.Individual);
vCard.setFullName("Jane Doe");
vCard.setName({ honorificPrefix: "Dr.", familyName: "Doe", givenName: "Jane"  });
vCard.setSocialProfile([
  { value: "https://instagram.com/", parameters: { type: "instagram" } },
  { value: "https://twitter.com/", parameters: { type: "twitter" } },
]);

vCard.toString();
```

```javascript
import generateVCard, { Kind, PropertyName } from "generate-vcard";

const vCard = generateVCard([
  { property: PropertyName.kind, value: Kind.Individual },
  { property: PropertyName.fullName, value: "Jane Doe" },
  { 
    property: Property.name,
    value: { honorificPrefix: "Dr.", familyName: "Doe", givenName: "Jane"  }
  },
  {
    property: Property.socialProfile,
    value: "https://instagram.com/",
    parameters: { type: "instagram" }
  },
  {
    property: Property.socialProfile,
    value: "https://twitter.com/",
    parameters: { type: "twitter" }
  }
]);
```


CHANGELOG.md
# [1.5.0](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.4.1...v1.5.0) (2022-08-03)


### Features

* fold lines to 75 octets ([b283daa](https://github.com/1nVitr0/lib-generate-vcard/commit/b283daac584ed53752718ef2a92f25038b2a579d))

## [1.4.1](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.4.0...v1.4.1) (2022-07-31)


### Bug Fixes

* export draft properties in experimental ([2e7f629](https://github.com/1nVitr0/lib-generate-vcard/commit/2e7f6297074b51e5af0f0b5836e61c18736dce4e))

# [1.4.0](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.3.0...v1.4.0) (2022-07-31)


### Features

* add common X-Name properties and parameters ([c5f7ff6](https://github.com/1nVitr0/lib-generate-vcard/commit/c5f7ff644923107ff266bfa5c4a2a71e67b31fab))

# [1.3.0](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.2.1...v1.3.0) (2022-07-28)


### Features

* implement extensions by OMA and CAB (RFC 6715) ([5eb4f12](https://github.com/1nVitr0/lib-generate-vcard/commit/5eb4f127c03c7893a68bff487f31e80af3e970f0))
* implement extensions by RDAP (RFC 8605) ([9e68d4c](https://github.com/1nVitr0/lib-generate-vcard/commit/9e68d4cb00bb203f32f6edab0e1f05fba21a6897))
* implement group parameter ([f8e1c12](https://github.com/1nVitr0/lib-generate-vcard/commit/f8e1c126253e1535eb46e22cf2e817c444fb2be4))

## [1.2.1](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.2.0...v1.2.1) (2022-07-28)


### Bug Fixes

* add package description ([ccddc62](https://github.com/1nVitr0/lib-generate-vcard/commit/ccddc627500a255ad7608e70395e59ca1c461493))

# [1.2.0](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.1.0...v1.2.0) (2022-07-28)


### Features

* allow providing parameters in VCard setters ([f62c7da](https://github.com/1nVitr0/lib-generate-vcard/commit/f62c7daf52dca3dcb8c6c4a0a366e6848732dce6))
* export Property values and parameters separately ([292af6c](https://github.com/1nVitr0/lib-generate-vcard/commit/292af6caa0abac3b5cae1cd4fb5a0e043327c0d8))
* support birth / death additions (RFC 6474) ([c90f6f6](https://github.com/1nVitr0/lib-generate-vcard/commit/c90f6f68b85e3ff937ed6062d2a7cd113f052daf))

# [1.1.0](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.0.4...v1.1.0) (2022-07-26)


### Features

* implement client pid map ([ea99237](https://github.com/1nVitr0/lib-generate-vcard/commit/ea99237d8d1e1de682fc8ed473f0766ae92722f6))

## [1.0.4](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.0.3...v1.0.4) (2022-07-26)


### Bug Fixes

* update token for automatic release ([670f205](https://github.com/1nVitr0/lib-generate-vcard/commit/670f2057538266155bc5c1a17160ae9eb0553cd0))

## [1.0.3](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.0.2...v1.0.3) (2022-07-26)


### Bug Fixes

* replace token for automatic release ([0f14fe6](https://github.com/1nVitr0/lib-generate-vcard/commit/0f14fe62e80c99a94f7270ca862db5e051d65875))

## [1.0.2](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.0.1...v1.0.2) (2022-07-26)


### Bug Fixes

* trigger release for ci test ([c52d2f3](https://github.com/1nVitr0/lib-generate-vcard/commit/c52d2f391a2cc579e573cb275579a1b00c512b95))

## [1.0.1](https://github.com/1nVitr0/lib-generate-vcard/compare/v1.0.0...v1.0.1) (2022-07-25)


### Bug Fixes

* add typedo documentation ([413797f](https://github.com/1nVitr0/lib-generate-vcard/commit/413797f4c8efaa5463c23c45e68f91a3e9d9e693))

# 1.0.0 (2022-07-25)


### Bug Fixes

* export all required types and funcitons ([fc48e84](https://github.com/1nVitr0/lib-generate-vcard/commit/fc48e8498121172b2c84a3ebd3ba144f40336f87))
* full test coverage ([2239bf9](https://github.com/1nVitr0/lib-generate-vcard/commit/2239bf91258ccba6f0b2cd4f22b4f5a37068c803))
* restructure typings ([3626d97](https://github.com/1nVitr0/lib-generate-vcard/commit/3626d97a1e4c8116bd9689782789c1e15637863d))


### Features

* add class based vcard generation ([25c23cc](https://github.com/1nVitr0/lib-generate-vcard/commit/25c23cc75910083fd97d8d24cdf3fe170d8165da))


### Performance Improvements

* reduce timestamp complexity ([cd7000e](https://github.com/1nVitr0/lib-generate-vcard/commit/cd7000e7929be93b725d6847092bad32c8c23a19))


src/index.ts
/**
 * A simple library for vCard generation compatible with [RFC 6350](https://tools.ietf.org/html/rfc6350),
 * [RFC 6474](https://tools.ietf.org/html/rfc6474). [RFC 6715](https://tools.ietf.org/html/rfc6715)
 * and [RFC 8605](https://tools.ietf.org/html/rfc8605).
 *
 * @example Functional / Object based { @link generateVCard }
 * ```ts
 * import generateVCard, { Kind } from "generate-vcard";
 *
 * const vCard = generateVCard({
 *   kind: Kind.Individual,
 *   fullName: "Jane Doe",
 * });
 * ```
 *
 * @example Functional / Array based { @link generateVCard }
 * ```ts
 * import generateVCard, { Kind, PropertyName } from "generate-vcard";
 *
 * const vCard = generateVCard([
 *   { property: PropertyName.kind, value: Kind.Individual },
 *   { property: PropertyName.fullName, value: "Jane Doe" },
 * ]);
 * ```
 *
 * @example Class based { @link VCard }
 * ```ts
 * import { VCard } from "generate-vcard";
 *
 * const vCard = new VCard(Kind.Individual);
 * vCard.setFullName("Jane Doe");
 *
 * vCard.toString();
 * ```
 *
 * @module generate-vcard
 */
import VCard from "./model/classes/VCard";
import { generateVCard } from "./generate/vCard";

export * from "./model/datatypes";
export * from "./model/properties";
export * from "./model/propertyDictionaries";
export * from "./model/propertyNames";
export * from "./model/propertyParameters";
export * from "./model/propertyValues";
export * from "./model/parameters";
export * from "./model/parameterNames";
export * from "./model/vCard";

export * from "./helpers/property";

export { VCard };
export default generateVCard;
